### PR TITLE
following_users

### PR DESCRIPTION
**Pull request by weilingooi**
finish following users

### CODE REVIEW

**Comment by weilingooi** [comment]
ready

File: app/controllers/relationships_controller.rb [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+
+  def destroy
+    @user = Relationship.find_by(id: params[:id]).followed
+    current_user.unfollow @user
+    respond_to do |format|
+      format.html { redirect_to @user }
+      format.js
+    end
+  end
+end
```
Comment (ngocvt-0484) [inline app/controllers/relationships_controller.rb]: please add endline

File: app/helpers/application_helper.rb [hunk 2]
```diff
@@ -1,6 +1,6 @@
 module ApplicationHelper
   def full_title page_title
-      
+
```
Comment (ngocvt-0484) [inline app/helpers/application_helper.rb]: remove empty line

File: app/models/user.rb [hunk 3]
```diff
@@ -82,3 +98,4 @@ def create_activation_digest
     self.activation_digest = User.digest activation_token
   end
 end
+end
```
Comment (ngocvt-0484) [inline app/models/user.rb]: add endline

File: app/models/user.rb [hunk 4]
```diff
@@ -4,7 +4,12 @@ class User < ApplicationRecord
   UPDATE_USERS_PARAMS = %i(name password password_confirmation).freeze
   attr_accessor :remember_token, :activation_token, :reset_token
   has_many :microposts, dependent: :destroy
-  before_save :downcase_email
+  has_many :active_relationships, class_name: Relationship.name, foreign_key: :follower_id, dependent: :destroy
+  has_many :passive_relationships, class_name: Relationship.name, foreign_key: :followed_id, dependent: :destroy
+  has_many :following, through: :active_relationships, source: :followed
+  has_many :followers, through: :passive_relationships, source: :follower
+  
+  before_save { email.downcase! }
```
Comment (ngocvt-0484) [inline app/models/user.rb]: why you remove line before_save :downcase_email, why don't use it?

File: app/models/user.rb [hunk 5]
```diff
@@ -4,7 +4,12 @@ class User < ApplicationRecord
...
   before_create :create_activation_digest
 
   before_save{self.email = email.downcase}
```
Comment (ngocvt-0484) [inline app/models/user.rb]: duplicate

File: app/views/shared/_stats.html.erb [hunk 6]
```diff
@@ -0,0 +1,15 @@
...
+    </strong>
+    <%= t "follow.following_title" %>
+  <% end %>
+  <%= link_to followers_path @user do %>
+    <strong id="followers" class="stat">
+      <%= @user.followers.count %>
+    </strong>
+    <%= t "follow.follower_title" %>
+  <% end %>
+</div>
```
Comment (ngocvt-0484) [inline app/views/shared/_stats.html.erb]: add endline

File: app/views/users/_user.html.erb [hunk 7]
```diff
@@ -0,0 +1,8 @@
+<li>
+  <%= gravatar_for user %>
+  <%= link_to user.name, user %>
+  <br>
+  <% if current_user.admin? %>
+    <%= link_to t(".delete"), user, method: :delete, data: {confirm: t(".confirm")}, class: "btn" %>
+  <% end %>
+</li>
```
Comment (ngocvt-0484) [inline app/views/users/_user.html.erb]: add endline

File: app/views/users/show_follow.html.erb [hunk 8]
```diff
@@ -0,0 +1,31 @@
...
+  <div class="col-md-8">
+    <h3><%= @title %></h3>
+    <% if @users.any? %>
+      <ul class="users follow">
+        <%= render @users %>
+      </ul>
+      <%= paginate @users, theme: "twitter-bootstrap-3" %>
+    <% end %>
+  </div>
+</div>
```
Comment (ngocvt-0484) [inline app/views/users/show_follow.html.erb]: add endline

File: config/routes.rb [hunk 9]
```diff
@@ -9,9 +9,18 @@
     get "/login", to: "sessions#new"
     post "/login", to: "sessions#create"
     delete "/logout", to: "sessions#destroy"
-    resources :users
+    get "activation/:token", to: "sessions#edit", as: "activation"
+    get ":id/following", to: "following#index", as: "following"
```
Comment (ngocvt-0484) [inline config/routes.rb]: it's  still work ok when remove this line right?

File: app/models/user.rb (hunk 4 above)
Comment (weilingooi) [reply app/models/user.rb]: error occured when i using before_save :downcase_email

File: config/routes.rb (hunk 9 above)
Comment (weilingooi) [reply config/routes.rb]: I put it below but not remove it...

**Comment by weilingooi** [comment]
ready

File: app/controllers/relationships_controller.rb [hunk 10]
```diff
@@ -0,0 +1,29 @@
...
+    @user = User.find_by id: params[:followed_id]
+    current_user.follow @user
+    respond_to do |format|
+      format.html { redirect_to @user }
+      format.js
+    end
+  end
+
+  def destroy
+    @user = Relationship.find_by(id: params[:id]).followed
```
Comment (manhnt-0960) [inline app/controllers/relationships_controller.rb]: please handle case `Relationship.find_by(id: params[:id])` return nil

File: app/controllers/users_controller.rb [hunk 11]
```diff
@@ -90,4 +93,5 @@ def is_admin
     flash[:warning] = t ".incorrect_user"
     redirect_to users_path
   end
+
```
Comment (manhnt-0960) [inline app/controllers/users_controller.rb]: let's remove empty line

File: app/assets/stylesheets/custom.scss [hunk 12]
```diff
@@ -200,3 +200,43 @@ span.picture {
...
+  a {
+    float: left;
+    padding: 0 10px;
+    border-left: 1px solid $logo_white;
+    color: gray;
+    &:first-child {
+      padding-left: 0;
+      border: 0;
+    }
+    &:hover {
```
Comment (chungpv-1008) [inline app/assets/stylesheets/custom.scss]:
```suggestion
    color: gray;
 
    &:first-child {
      padding-left: 0;
      border: 0;
    }
 
    &:hover {
```

File: app/assets/stylesheets/custom.scss [hunk 13]
```diff
@@ -200,3 +200,43 @@ span.picture {
...
+  strong {
+    display: block;
+  }
+}
+
+.user_avatars {
+  overflow: auto;
+  margin-top: 10px;
+  .gravatar {
+    margin: 1px 1px;
```
Comment (chungpv-1008) [inline app/assets/stylesheets/custom.scss]:
```suggestion
  margin-top: 10px;

  .gravatar {
    margin: 1px 1px;
```

File: config/routes.rb (hunk 9 above)
Comment (manhnt-0960) [reply config/routes.rb]: let's remove it

File: app/controllers/relationships_controller.rb [hunk 14]
```diff
@@ -0,0 +1,29 @@
...
+  def index
+    @user = User.find_by id: params[:id]
+    redirect_to root_path unless @user
+    @title = t ".#{params[:type]}"
+    @users = @user.send(params[:type]).page params[:page]
+    render "users/show_follow"
+  end
+
+  def create
+    @user = User.find_by id: params[:followed_id]
```
Comment (chungpv-1008) [inline app/controllers/relationships_controller.rb:14]: add before_action for action 'index, create'
```
def find_user
    @user = User.find_by id: params[:id] || params[:followed_id]
    return if @user.present?
    
    flash[:danger] = t ".invalid_user"
    redirect_to root_path
  end
```

File: app/controllers/users_controller.rb [hunk 15]
```diff
@@ -6,6 +6,8 @@ class UsersController < ApplicationController
     
   def show
     @microposts = @user.microposts.paginate page: params[:page]
+    @follow = current_user.active_relationships.build
+    @unfollow = current_user.active_relationships.find_by followed_id: @user.id
```
Comment (chungpv-1008) [inline app/controllers/users_controller.rb]:
```suggestion
    @unfollow = current_user.following
```

File: app/models/user.rb [hunk 16]
```diff
@@ -4,10 +4,13 @@ class User < ApplicationRecord
   UPDATE_USERS_PARAMS = %i(name password password_confirmation).freeze
   attr_accessor :remember_token, :activation_token, :reset_token
   has_many :microposts, dependent: :destroy
+  has_many :active_relationships, class_name: Relationship.name, foreign_key: :follower_id, dependent: :destroy
+  has_many :passive_relationships, class_name: Relationship.name, foreign_key: :followed_id, dependent: :destroy
+  has_many :following, through: :active_relationships, source: :followed
+  has_many :followers, through: :passive_relationships, source: :follower
```
Comment (chungpv-1008) [inline app/models/user.rb]:
```suggestion
  has_many :followers, through: :passive_relationships
```

File: app/views/relationships/create.js.erb [hunk 17]
```diff
@@ -0,0 +1,6 @@
+<% if @error.present? %>
+  alert("<%= @error %>")
+<% else %>
+  $("#follow_form").html("<%= j render "users/unfollow" %>");
+  $("#followers").html("<%= @user.followers.count %>");
```
Comment (chungpv-1008) [inline app/views/relationships/create.js.erb]: count -> size

File: app/views/relationships/destroy.js.erb [hunk 18]
```diff
@@ -0,0 +1,6 @@
+<% if @error.present? %>
+  alert("<%= @error %>")
+<% else %>
+  $("#follow_form").html("<%= j render "users/follow" %>");
+  $("#followers").html("<%= @user.followers.count %>");
```
Comment (chungpv-1008) [inline app/views/relationships/destroy.js.erb]: similar

File: app/views/shared/_stats.html.erb [hunk 19]
```diff
@@ -0,0 +1,15 @@
...
+    <strong id="following" class="stat">
+      <%= @user.following.count %>
+    </strong>
+    <%= t "follow.following_title" %>
+  <% end %>
+  <%= link_to followers_path @user do %>
+    <strong id="followers" class="stat">
+      <%= @user.followers.count %>
+    </strong>
+    <%= t "follow.follower_title" %>
```
Comment (chungpv-1008) [inline app/views/shared/_stats.html.erb]:
```suggestion
  <%= link_to following_path(@user) do %>
    <strong id="following" class="stat">
      <%= @user.following.size %>
    </strong>
    <%= t "follow.following_title" %>
  <% end %>
  <%= link_to followers_path(@user) do %>
    <strong id="followers" class="stat">
      <%= @user.followers.size %>
    </strong>
    <%= t "follow.follower_title" %>
```

File: app/views/users/show_follow.html.erb [hunk 20]
```diff
@@ -0,0 +1,31 @@
+<% provide :title, @title %>
+<div class="row">
+  <aside class="col-md-4">
+    <section class="user_info">
+      <%= gravatar_for @user %>
+      <h1><%= @user.name %></h1>
+      <span><%= link_to t "follow.view_profile", @user %></span>
```
Comment (chungpv-1008) [inline app/views/users/show_follow.html.erb]:
```suggestion
      <span><%= link_to t("follow.view_profile"), @user %></span>
```

File: app/views/users/show_follow.html.erb [hunk 21]
```diff
@@ -0,0 +1,31 @@
...
+      <span><b><%= t "microposts.title" %></b>
+        <%= @user.micro_posts.count %></span>
```
Comment (chungpv-1008) [inline app/views/users/show_follow.html.erb]:
```suggestion
        <%= @user.micro_posts.size %></span>
```

**Comment by weilingooi** [comment]
ready

File: app/controllers/relationships_controller.rb (hunk 14 above)
Comment (ngocvt-0484) [reply app/controllers/relationships_controller.rb:14]: commented

**Comment by ngocvt-0484** [comment]
commented

**Comment by weilingooi** [comment]
ready

File: app/controllers/relationships_controller.rb [hunk 22]
```diff
@@ -0,0 +1,34 @@
...
+    relationship = Relationship.find_by id: params[:id]
+    if relationship.present?
+      @user = relationship.followed
+      current_user.unfollow @user
+    else
+      @error = t "follow.action_fail"
+    end
+    respond_to do |format|
+      format.js
+    end
```
Comment (chungpv-1008) [inline app/controllers/relationships_controller.rb]:
```suggestion
    respond_to :js
```

**Review by chungpv-1008 — COMMENTED** [review]
please fix comment, then you merge yourself

**Pull request merged**